restli-client/src/test/java/com/linkedin/restli/client/TestExecutionGroup.java

```diff
@@ -0,0 +1,230 @@
+    _parSeqUnitTestHelper.setUp();
+    _engine = _parSeqUnitTestHelper.getEngine();
+  }
+
+  @BeforeMethod
+  public void doBeforeMethod()
+  {
+    eg = new ExecutionGroup(_engine);
+    client1 = mock(MockBatchableResource.class);
+    when(client1.get(any())).thenReturn("1");
+    task1 = Task.callable(() -> {
+      client1.get(1L);
+      return null;
+    });
+  }
+
+  @Test
+  public void testAddToGroupAndExecute_SingleGroup() throws Exception
+  {
+    eg.addTaskByFluentClient(client1, task1);
+    eg.execute();
+    Assert.assertEquals(eg.getClientToTaskListMap().size(), 0);
+    awaitAllTasks(task1);
+    verify(client1, times(1)).get(any());
+  }
+
+  @Test
+  public void testAddToGroupAndExecute_MultipleGroup() throws Exception
+  {
+    // Task will be called twice in two groups
+    Task mockTask = mock(Task.class);
+    ExecutionGroup eg2 = new ExecutionGroup(_engine);
+    eg.addTaskByFluentClient(client1, mockTask);
+    eg2.addTaskByFluentClient(client1, mockTask);
+    eg.execute();
+    eg2.execute();
+    Thread.sleep(200); // wait for engine to execute
+    verify(mockTask, times(2)).contextRun(any(), any(), any());
+  }
+
+  @Test
+  public void testBatching() throws Exception
+  {
+    // Test at least tasks from single client should be batched
+    String value1 = "TASK1";
+    String value2 = "TASK2";
+    _resourceClient = mock(MockBatchableResource.class);
+    // Create a strategy that will run the batch on every 2 requests;
+    MockBatchingStrategy _mockBatchingStrategy = new MockBatchingStrategy(_resourceClient, 2);
+    Task<String> t1 = _mockBatchingStrategy.batchable(1L);
+    Task<String> t2 = _mockBatchingStrategy.batchable(2L);
+    eg.addTaskByFluentClient(client1, t1, t2);
+    eg.execute();
+    when(_resourceClient.batchGet(any())).thenReturn(new HashMap<Long, String>()
+    {{
+      put(1L, value1);
+      put(2L, value2);
+    }});
+    awaitAllTasks(t1, t2);
+    verify(_resourceClient, never()).get(any());
+    verify(_resourceClient, times(1)).batchGet(any());
+  }
+
+  @Test
+  public void testCallableExecution() throws Exception
+  {
+    MockBatchableResource client = new MockBatchableResource();
+    eg.batchOn(() -> {
+      Assert.assertTrue(client.validateExecutionGroupFromContext(eg));
+    }, client);
+  }
+
+  @Test
+  public void testCallableExecution_Nested() throws Exception
+  {
+    MockBatchableResource client = new MockBatchableResource();
+    eg.batchOn(() -> {
+      Assert.assertTrue(client.validateExecutionGroupFromContext(eg)); //outer - eg
+      ExecutionGroup eg2 = new ExecutionGroup(_engine);
+      try {
+        eg2.batchOn(() -> {
+          Assert.assertTrue(client.validateExecutionGroupFromContext(eg2)); // inner - eg2
+        }, client);
+      } catch (Exception ignored) {
+      }
+      Assert.assertTrue(client.validateExecutionGroupFromContext(eg)); // outer -eg
+    }, client);
+  }
+
+  @Test public void testExecuteOnlyOnce() throws Exception
```

Comment:
nit: Use @Test(expectedExceptions, expectedExceptionsMessageRegExp) 
Avoids the try catch and explict fail message.

Reply:
I thought about that. But that doesn't catch fail message.

So why are you against catching the fail message?

Reply:
btw, the error here is actually CompletionErroor.

Reply:
The code is actually catching IllegalStateException?

if you did
```
@Test(expectedExceptions={IllegalstateException.class}, expectedExceptionsMessageRegExp=ExecutionGroup.MULTIPLE_EXECUTION_ERROR)
```
Then the test would behave as what you have today. It would fail if the exception was not thrown or if a wrong exception was thrown.

Reply:
It is not that important, as I said, it is a  nit as this is a slightly better way to define this expectation (easier to read).